src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,645 @@
+        {
+            public CSharpAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+                : base(semanticModel, cancellationToken)
+            {
+            }
+
+            protected override string Title => CSharpFeaturesResources.Convert_linq_query_to_foreach;
+
+            protected override bool TryConvert(QueryExpressionSyntax source, out DocumentUpdate documentUpdate)
+            {
+                if (_semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any())
+                {
+                    documentUpdate = default;
+                    return false;
+                }
+
+                if (source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                    SyntaxKind.SingleLineCommentTrivia,
+                    SyntaxKind.MultiLineCommentTrivia,
+                    SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                    source.ContainsDirectives))
+                {
+                    documentUpdate = default;
+                    return false;
+                }
+
+                // TODO add a semantic check
+                // validate that all clauses are translated into methods with signatures corresponding to code generated by the refactoring.
+
+                switch (source.Parent.Kind())
+                {
+                    case SyntaxKind.ReturnStatement:
+                        return TryConvertInReturnStatement(source, out documentUpdate);
+                    case SyntaxKind.ForEachStatement:
+                        return TryConvertInForEach(source, out documentUpdate);
+                    case SyntaxKind.EqualsValueClause:
+                    case SyntaxKind.SimpleAssignmentExpression:
+                    case SyntaxKind.MultiplyAssignmentExpression:
+                        return TryReplaceWithLocalFunction(source, out documentUpdate);
+                    case SyntaxKind.ParenthesizedExpression:
+                        return TryConvertInParenthesizedExpression(source, out documentUpdate);
+                    default:
+                        documentUpdate = default;
+                        return false;
+                }
+            }
+
+            private bool TryConvertInParenthesizedExpression(QueryExpressionSyntax source, out DocumentUpdate documentUpdate)
+            {
+                // TODO any other cases?
+                var parenthesizedExpression = (ParenthesizedExpressionSyntax)source.Parent;
+                if (parenthesizedExpression.Parent.Kind() == SyntaxKind.SimpleMemberAccessExpression)
+                {
+                    var memberAccessExpression = (MemberAccessExpressionSyntax)parenthesizedExpression.Parent;
+
+                    Func<ExpressionSyntax, ExpressionSyntax, ExpressionSyntax> insideLoopExpression;
+                    ExpressionSyntax initializer;
+                    string variableNamePrefix;
```

Comment:
really hard to read code that declares variables like this, but then sets them through some codepaths.  Consider a pattern where you call something that comptues and returns these values, then use them if you got hem back.